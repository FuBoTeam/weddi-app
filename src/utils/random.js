const random = (a = 1, b = 0) => {
  const lower = Math.ceil(Math.min(a, b));
  const upper = Math.floor(Math.max(a, b));
  return Math.floor(lower + Math.random() * (upper - lower + 1));
};

export const combinationList = (list, k) => {
  const result = [];
  /* select k elements n-k <= i < n */
  for (let i = list.length - k; i < list.length; i++) {
    /* A random integer such that 0 <= j < i */
    const j = random(i);
    /* Select element j if element j is not selected yet */
    /* Otherwise select element i */
    if (result.includes(list[j])) {
      result.push(list[i]);
    } else {
      result.push(list[j]);
    }
  }
  return result;
};

export const permutationList = list => {
  for (let i = 0; i <= list.length - 2; i++) {
    /* A random integer such that i <= j < n */
    const j = random(i, list.length - 1);
    /* Swap the randomly picked element with list[i] */
    [list[i], list[j]] = [list[j], list[i]];
  }
  return list;
};
