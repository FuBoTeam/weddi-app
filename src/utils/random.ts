import random from 'lodash/random';

export function combinationList<T>(list: T[], k: number): T[] {
  if (list.length < k) {
    return list;
  }
  const result: T[] = [];
  /* select k elements n-k <= i < n */
  for (let i = list.length - k; i < list.length; i++) {
    /* A random integer such that 0 <= j < i */
    const j = random(i);
    /* Select element j if element j is not selected yet */
    /* Otherwise select element i */
    if (result.includes(list[j])) {
      result.push(list[i]);
    } else {
      result.push(list[j]);
    }
  }
  return result;
}

export function permutationList<T>(list: T[]): T[] {
  if (list.length < 2) {
    return list;
  }
  for (let i = 0; i <= list.length - 2; i++) {
    /* A random integer such that i <= j < n */
    const j = random(i, list.length - 1);
    /* Swap the randomly picked element with list[i] */
    [list[i], list[j]] = [list[j], list[i]];
  }
  return list;
}
