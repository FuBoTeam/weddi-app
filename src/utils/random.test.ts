import { combinationList, permutationList } from './random';

describe('random', () => {
  beforeEach(() => {
    jest.spyOn(global.Math, 'random').mockReturnValue(0.123456789);
  });

  afterEach(() => {
    jest.spyOn(global.Math, 'random').mockRestore();
  });
  describe('combinationList', () => {
    describe('when given a list with length n = 10', () => {
      const n = 10;
      const list = Array.from({length: n}, (_, i) => i);
      describe.each([
        [1],
        [2],
        [5],
        [9],
        [10],
      ])('and given a valid k = %d', (k) => {
        it('should return a list with length k', () => {
          expect(combinationList(list, k)).toHaveLength(k);
        });
        it('should match the snapshot', () => {
          expect(combinationList(list, k)).toMatchSnapshot();
        });
      });
    });
  });
  describe('permutationList', () => {
    describe('when given a list with length n = 10', () => {
      const n = 10;
      const list = Array.from({length: n}, (_, i) => i);
      it('should return a list with length n', () => {
        expect(permutationList(list)).toHaveLength(n);
      });
      it('should match the snapshot', () => {
        expect(permutationList(list)).toMatchSnapshot();
      });
    });
  });
});
