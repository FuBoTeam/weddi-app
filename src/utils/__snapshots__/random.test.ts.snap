// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`random combinationList when given a list with length n = 10 and given a valid k = 1 should match the snapshot 1`] = `
Array [
  1,
]
`;

exports[`random combinationList when given a list with length n = 10 and given a valid k = 2 should match the snapshot 1`] = `
Array [
  1,
  9,
]
`;

exports[`random combinationList when given a list with length n = 10 and given a valid k = 5 should match the snapshot 1`] = `
Array [
  0,
  6,
  7,
  1,
  9,
]
`;

exports[`random combinationList when given a list with length n = 10 and given a valid k = 9 should match the snapshot 1`] = `
Array [
  0,
  2,
  3,
  4,
  5,
  6,
  7,
  1,
  9,
]
`;

exports[`random combinationList when given a list with length n = 10 and given a valid k = 10 should match the snapshot 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
]
`;

exports[`random permutationList when given a list with length n = 10 should match the snapshot 1`] = `
Array [
  2,
  0,
  1,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
]
`;
